src/nnbench/runner.py

```diff
@@ -187,7 +187,7 @@ def run(
         path_or_module: str | os.PathLike[str],
         params: dict[str, Any] | Parameters | None = None,
         tags: tuple[str, ...] = (),
-        context: Sequence[ContextProvider] = (),
+        context: Sequence[ContextProvider | dict[str, Any] | Context] = (),
```

Comment:
A sequence of contexts? I'm not sure that's correct. We take one `Context` object, and add/update it throughout our benchmarks.

Can you elaborate on why this was necessary?

Reply:
I wanted to pass information to the context parameter of the runner. Specifically the name of the validation data set that I used (there are multiple versions of the CoNLL dataset).

As the `Context.update` also handles `ContextProvider | dict[str, Any] | "Context"`, I thought it makes sense that the runner can handle all these types as well since the runner uses `Context.update()` to handle the provided context.

Reply:
Good thinking. I'm not a fan of this typing, which is because the `Context.update` typing is too permissive.

Since the context is supposed to group different value sets on top-level keys (e.g. git info under "git"), I think the following should be true:

- `Context.update` merges two `Context` objects, so "other" needs to be a `Context`.
- `Context.add(provider)` inserts a context provider's result into the context under a single top-level context key.
- We drop raw dict support entirely, outside of `Context.make()` and the constructor.

Then, the typing of `context` in `BenchmarkRunner.run()` becomes `Context | Sequence[ContextProvider]` - the former if you want to supply your own hand-built context, the latter if you want it assembled from atomic providers. How does that sound?